pgtoolkit/ctl.py

```diff
@@ -0,0 +1,248 @@
+"""
+.. currentmodule:: pgtoolkit.ctl
+
+API Reference
+-------------
+
+.. autoclass:: PGCtl
+    :members:
+.. autoclass:: Status
+    :members:
+.. autofunction:: run_command
+.. autoclass:: CommandRunner
+    :members: __call__
+"""
+
+import enum
+import shutil
+import subprocess
+import sys
+from pathlib import Path
+from typing import Any, List, Mapping, Optional, Sequence, Union, TYPE_CHECKING
+from typing_extensions import Literal, Protocol
+
+if TYPE_CHECKING:
+    CompletedProcess = subprocess.CompletedProcess[str]
+else:
+    CompletedProcess = subprocess.CompletedProcess
+
+PY37 = sys.version_info[:1] >= (3, 7)
+
+
+class CommandRunner(Protocol):
+    """Protocol for `run_command` callable parameter of :class:`PGCtl`.
+
+    The `text` mode, as defined in :mod:`subprocess`, must be used in
+    implementations.
+
+    Keyword arguments are expected to match that of :func:`subprocess.run`.
+    """
+
+    def __call__(
+        self,
+        args: Sequence[str],
+        *,
+        capture_output: bool = False,
+        check: bool = False,
+        **kwargs: Any,
+    ) -> CompletedProcess:
+        ...
+
+
+def run_command(
+    args: Sequence[str],
+    *,
+    capture_output: bool = False,
+    check: bool = False,
+    **kwargs: Any,
+) -> CompletedProcess:
+    """Default :class:`CommandRunner` implementation for :class:`PGCtl` using
+    :func:`subprocess.run`.
+    """
+    if PY37:
+        kwargs["capture_output"] = capture_output
+    elif capture_output:
+        if "stdout" in kwargs or "stderr" in kwargs:
+            raise ValueError(
+                "stdout and stderr arguments may not be used with capture_output"
+            )
+        kwargs["stdout"] = kwargs["stderr"] = subprocess.PIPE
+
+    return subprocess.run(
+        args,
+        check=check,
+        universal_newlines=True,
+        **kwargs,
+    )
+
+
+def _args_to_opts(args: Mapping[str, Union[str, Literal[True]]]) -> List[str]:
+    options = []
+    for name, value in sorted(args.items()):
```

Comment:
Is this `sorted` required?

Reply:
Not exactly required but I feel it's better to have this because it makes the order of options consistent with the order in which they are specified in function call and also because it simplifies testing.